### PR TITLE
Irrlicht cleanups (mostly getting rid of `core::array`)

### PR DESCRIPTION
**Pull request by sfan5**
to be clear: this is about half of the places `core::array` was still used

## To do

This PR is Ready for Review.

## How to test

1. run game
2. check that things still work (esp. rendering & models)

### CODE REVIEW

**Review by appgurueu — COMMENTED** [review]
Thanks for ripping out `core::array`. I have mostly minor comments. Tested by playing on CTF, some MTG, and some devtest, no issues related to this PR found (but I did find an unrelated issue: #14985 :sweat_smile:).

File: irr/src/CAttributes.cpp [hunk 1]
```diff
@@ -12,51 +12,32 @@ namespace irr
...
 {
-	for (u32 i = 0; i < Attributes.size(); ++i)
-		if (Attributes[i]->Name == attributeName)
-			return Attributes[i];
-
-	return 0;
+	auto it = Attributes.find(attributeName);
+	if (it == Attributes.end())
+		return nullptr;
+	return const_cast<IAttribute*>(it->second); // who cares
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/src/CAttributes.cpp:40]: I was too lazy to add a const variant (or better: inline it entirely)

File: irr/src/CAttributes.cpp [hunk 2]
```diff
@@ -66,7 +47,7 @@ void CAttributes::setAttribute(const c8 *attributeName, bool value)
 	if (att)
 		att->setBool(value);
 	else {
-		Attributes.push_back(new CBoolAttribute(attributeName, value));
+		Attributes[attributeName] = new CBoolAttribute(attributeName, value);
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/src/CAttributes.cpp:50]: Not very but I didn't want to put too much work into such an irrelevant part.

File: irr/src/CAttributes.cpp [hunk 3]
```diff
@@ -158,91 +123,10 @@ E_ATTRIBUTE_TYPE CAttributes::getAttributeType(const c8 *attributeName) const
...
-	if ((u32)index < Attributes.size())
-		Attributes[index]->setInt(value);
-}
-
-//! Sets a attribute as float value
-void CAttributes::setAttribute(s32 index, f32 value)
-{
-	if ((u32)index < Attributes.size())
-		Attributes[index]->setFloat(value);
+	return !!getAttributeP(attributeName);
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/src/CAttributes.cpp:129]: that's longer

File: irr/src/CBillboardSceneNode.cpp [hunk 4]
```diff
@@ -114,7 +114,7 @@ void CBillboardSceneNode::updateMesh(const irr::scene::ICameraSceneNode *camera)
 
 	view *= -1.0f;
 
-	core::array<video::S3DVertex> &vertices = Buffer->Vertices;
+	auto *vertices = Buffer->Vertices.data();
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/src/CBillboardSceneNode.cpp:117]: no particular reason. does it make a difference?

File: irr/src/CNullDriver.cpp [hunk 5]
```diff
@@ -361,7 +360,7 @@ ITexture *CNullDriver::addTextureCubemap(const io::path &name, IImage *imagePosX
 
 	ITexture *t = 0;
 
-	core::array<IImage *> imageArray(6);
+	std::vector<IImage*> imageArray;
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/src/CNullDriver.cpp:363]: I was too lazy to make more changes.

File: irr/src/OpenGL/MaterialRenderer.cpp [hunk 6]
```diff
@@ -262,35 +262,30 @@ bool COpenGL3MaterialRenderer::linkProgram()
 
 		// seems that some implementations use an extra null terminator.
 		++maxlen;
-		c8 *buf = new c8[maxlen];
+		std::vector<c8> buf(maxlen);
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/src/OpenGL/MaterialRenderer.cpp:265]: I think this is irrelevant. Contrary actually I think a vector could potentially be stack-allocated.

File: irr/src/CBillboardSceneNode.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (appgurueu) [reply irr/src/CBillboardSceneNode.cpp:117]: Not a significant one, but vector access could e.g. get bounds checks if something like [this](https://gcc.gnu.org/onlinedocs/libstdc++/manual/debug_mode.html) is used. Primarily a matter of taste.

**Review by appgurueu — APPROVED** [review]

**Pull request merged**